From db976c59d6184d0235ca0645fc9fe93d11110079 Mon Sep 17 00:00:00 2001
From: Jack Ren <jack.ren@intel.com>
Date: Thu, 2 Dec 2010 01:07:58 +0800
Subject: [PORT FROM MAIN] fastboot: fix the failure when the fastboot is not first function in composite

BZ: 70148

Change-Id: Ieee8db3770e844bfb1c35c9c70024d0ac167ba26
Orig-Change-Id: I5a3d57ca49e8cf6fed08fd6504a2ff67088a1fe1
Signed-off-by: Jack Ren <jack.ren@intel.com>
Reviewed-on: http://android.intel.com:8080/75780
Reviewed-by: Tinelli, Vincent <vincent.tinelli@intel.com>
Reviewed-by: Romieu, Benoit <benoit.romieu@intel.com>
Tested-by: Romieu, Benoit <benoit.romieu@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 fastboot/usb_linux.c |   18 ++++++++++++++----
 1 files changed, 14 insertions(+), 4 deletions(-)

diff --git a/fastboot/usb_linux.c b/fastboot/usb_linux.c
index b7a9ca3..15f2867 100644
--- a/fastboot/usb_linux.c
+++ b/fastboot/usb_linux.c
@@ -198,8 +198,13 @@ static int filter_usb_device(int fd, char *ptr, int len, int writable,
     }
 
     for(i = 0; i < cfg->bNumInterfaces; i++) {
-        if(check(ptr, len, USB_DT_INTERFACE, USB_DT_INTERFACE_SIZE))
-            return -1;
+        if(check(ptr, len, USB_DT_INTERFACE, USB_DT_INTERFACE_SIZE)) {
+            len -= ptr[0];
+            ptr += ptr[0];
+            i--;
+            continue;
+        }
+
         ifc = (void*) ptr;
         len -= ifc->bLength;
         ptr += ifc->bLength;
@@ -211,8 +216,13 @@ static int filter_usb_device(int fd, char *ptr, int len, int writable,
         info.ifc_protocol = ifc->bInterfaceProtocol;
 
         for(e = 0; e < ifc->bNumEndpoints; e++) {
-            if(check(ptr, len, USB_DT_ENDPOINT, USB_DT_ENDPOINT_SIZE))
-                return -1;
+            if(check(ptr, len, USB_DT_ENDPOINT, USB_DT_ENDPOINT_SIZE)) {
+            len -= ptr[0];
+            ptr += ptr[0];
+            e--;
+            continue;
+        }
+
             ept = (void*) ptr;
             len -= ept->bLength;
             ptr += ept->bLength;
-- 
1.7.4.1

